src/main/client/batch_read.c

```diff
@@ -0,0 +1,299 @@
+            Py_DECREF(py_batch_record);
+            break;
+        }
+
+        PyList_Append(data->py_results, py_batch_record);
+        Py_DECREF(py_batch_record);
+    }
+
+    PyGILState_Release(gstate);
+    return success;
+}
+
+PyObject *AerospikeClient_BatchRead(AerospikeClient *self, PyObject *args,
+                                    PyObject *kwds)
+{
+    PyObject *py_keys = NULL;
+    PyObject *py_bins = NULL;
+    PyObject *py_policy_batch = NULL;
+    static char *kwlist[] = {"keys", "bins", "policy", NULL};
+    if (PyArg_ParseTupleAndKeywords(args, kwds, "O|OO:batch_read", kwlist,
+                                    &py_keys, &py_bins,
+                                    &py_policy_batch) == false) {
+        return NULL;
+    }
+
+    as_error err;
+    as_error_init(&err);
+
+    // required arg so don't need to check for NULL
+    if (!PyList_Check(py_keys)) {
+        as_error_update(&err, AEROSPIKE_ERR_PARAM,
+                        "keys should be a list of aerospike key tuples");
+        goto CLEANUP1;
+    }
+
+    as_vector tmp_keys;
+    Py_ssize_t keys_size = PyList_Size(py_keys);
+    as_vector_init(&tmp_keys, sizeof(as_key), keys_size);
+    as_vector *tmp_keys_p = &tmp_keys;
+
+    if (!self || !self->as) {
+        as_error_update(&err, AEROSPIKE_ERR_PARAM, "Invalid aerospike object");
+        goto CLEANUP2;
+    }
+
+    if (!self->is_conn_16) {
+        as_error_update(&err, AEROSPIKE_ERR_CLUSTER,
+                        "No connection to aerospike cluster");
+        goto CLEANUP2;
+    }
+
+    uint64_t processed_key_count = 0;
+    for (int i = 0; i < keys_size; i++) {
+        PyObject *py_key = PyList_GetItem(py_keys, i);
+        as_key *tmp_key = (as_key *)as_vector_get(&tmp_keys, i);
+
+        Py_XINCREF(py_key);
+        if (!PyTuple_Check(py_key)) {
+            as_error_update(&err, AEROSPIKE_ERR_PARAM,
+                            "key should be an aerospike key tuple");
+            Py_XDECREF(py_key);
+            goto CLEANUP2;
+        }
+
+        pyobject_to_key(&err, py_key, tmp_key);
+        if (err.code != AEROSPIKE_OK) {
+            as_error_update(&err, AEROSPIKE_ERR_PARAM,
+                            "failed to convert key at index: %d", i);
+            goto CLEANUP2;
+        }
+
+        processed_key_count++;
+    }
+
+    as_batch batch;
+    as_batch_init(&batch, processed_key_count);
+    memcpy(batch.keys.entries, tmp_keys.list,
+           sizeof(as_key) * processed_key_count);
+
+    as_policy_batch policy_batch;
+    as_policy_batch *policy_batch_p = NULL;
+
+    // For expressions conversion.
+    as_exp batch_exp_list;
+    as_exp *batch_exp_list_p = NULL;
+
+    if (py_policy_batch) {
+        if (pyobject_to_policy_batch(
+                self, &err, py_policy_batch, &policy_batch, &policy_batch_p,
+                &self->as->config.policies.batch, &batch_exp_list,
+                &batch_exp_list_p) != AEROSPIKE_OK) {
+            goto CLEANUP3;
```

Comment:
I think all gotos after this skip freeing the expressions allocated here.

Reply:
Fixed